state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser/block_processor.rs

```diff
@@ -0,0 +1,625 @@
+use crate::electoral_systems::{
+	block_witnesser::{primitives::ChainProgressInner, state_machine::BWProcessorTypes},
+	state_machine::{
+		core::{Hook, IndexOf, Indexed, Validate},
+		state_machine2::StateMachine,
+	},
+};
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{pallet_prelude::TypeInfo, Deserialize, Serialize};
+use sp_std::{collections::btree_map::BTreeMap, fmt::Debug, marker::PhantomData, vec, vec::Vec};
+
+/// A helper trait for determining whether two objects are “inner‐equal.”
+/// This trait is intended for use with event types to provide a custom notion
+/// of equality. In our context, events may be produced via different Rules (such as pre‐witness
+/// and full witness), and we want to ensure that the “deposit witness” aspect of an event is
+/// compared when deduplicating events.
+pub trait InnerEquality {
+	fn inner_eq(&self, other: &Self) -> bool;
+}
```

Comment:
I've been thinking about this, and maybe there is a way to do this without this trait. If we split the event into two parts of "blockdata" and "deposit data", then we could compare the deposit data independently of the blockheight/hash to check which events we already emitted. The `Execute` hook would then take two parameters, i.e.
```rust
Hook<(Self::ChainBlockNumber, Self::Event), ()>
```